src/indexing_str.rs

```diff
@@ -41,6 +42,22 @@ unsafe impl Contiguous for Str {
 
 impl Str {
     pub fn slice<'a, 'b, 'i>(input: &'b Container<'i, &'a Self>, range: Range<'i>) -> &'b Self {
-        unsafe { &*(&input[range] as *const [u8] as *const Str) }
+        // NOTE(eddyb) following code is copied from `str::is_char_boundary`:
+        let valid_utf8_start = |bytes: &[u8]| {
+            match bytes.first() {
```

Comment:
Use `Option::filter`?

Reply:
For what?

Reply:
`bytes.first().filter(|&&b| (b as i8) >= -0x40).is_some()`

maybe it's better as is...

Reply:
Note the `NOTE` comment above :P.